### DISCUSSION TITLE
Official container image

### DISCUSSION THREAD

**Discussion by bdd** [bdd/runitor#104] [General]
Based on a user request in #95, I decided to try building a multi architecture image for runitor and publish on Docker Hub, using only Podman. (For reasons irrelevant to discussion, I don't want to rely on any Docker tooling).

For now [docker.io/runitor/runitor](https://hub.docker.com/r/runitor/runitor) has a Alpine based images of v1.2.0 release (and symbolic `latest`) for `amd64`, `arm64/v8`, `arm/v7`, and `arm/v6` architectures. The images offer no entrypoint as they are intended to be the base layer for your workload images.

As I mentioned in the issue thread, I was only able to test the ARM images under binfmt with QEMU emulation. So I have a favor to ask. If you run Linux containers on an ARM machine, please test and report.

I understand there can be demand for more traditional Linux distribution based images, rather than Alpine _(which uses musl, BusyBox, OpenRC instead of glibc, core utils, systemd)_. So I intend to add at least Debian based image as well. Currently I don't plan to be as comprehensive as Go's images where they support multiple Debian and Alpine releases.

If you have opinions, this is the right thread to chime in.

I'm going to tag two of the runitor container image maintainers, @markcaudill and @skynewz here to give them a chance to chime in as well if they are interested.

**Comment by markcaudill** [comment]
I also ran into the dilemma regarding to musl vs glibc and landed on supporting Alpine and Debian. It feels sub-optimal but I don't know a better strategy.

Anyway, I'm thrilled that this has been added to the official repo so I'll probably be marking my version as archived/deprecated in the near future.

ETA: I'm also a big fan of avoiding Docker tooling whenever possible. 👍🏻

**Reply by bdd** [reply]
A few things changed since I posted above.
- There are now 3 base images available through their tags. Alpine, Debian, and Ubuntu. `runitor:latest` uses `ubuntu:latest`. I made this choice in a mostly pseudo-scientific way: Ubuntu images had more pulls than Alpine and Debian.
- I added an ENTRYPOINT and a CMD to the image because previously I misunderstood inheritance rules (there is none), and having one gives some kind of smoke testing convenience, e.g. `podman run docker.io/runitor/runitor -uuid <...> sh -c 'hello, world!'`
- I dropped `arm/v6` support to have identical arch support across images. People with first generation RPi won't be able to run it but I doubt there are many.

### Tooling
Commit https://github.com/bdd/runitor/commit/7935b789fb4546d7b31124f1ada681d87a866b69 adds the `Containerfile` and the script I use to build multi-arch manifests and push them. You'd see it relies on signed binary builds, and verifies them. So traceability is not hindered. Contributions are most welcome.

### The issue with Docker, Inc.
I'm going to explore publishing to another OCI repository in addition to Docker Hub. That company sadly doesn't act very stable lately. I'm all ears if you have recommendations.
